### PR TITLE
[jvm-packages] add doc for xgboost4j-spark-gpu

### PR DESCRIPTION
**Pull request by wbo4958**
This PR adds the tutorial for how to use the XGBoost4j-Spark-GPU.

### CODE REVIEW

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,281 @@
+#############################################
+XGBoost4J-Spark-GPU Tutorial (version 1.6.0+)
+#############################################
+
+**XGBoost4J-Spark-GPU** is a project aiming to accelerate XGBoost distributed training on Spark from
+end to end by GPUs by leveraging the ``Spark-Rapids``.
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:6]:
```suggestion
end to end with GPUs by leveraging the ``Spark-Rapids`` project.
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+``Spark-Rapids`` is the RAPIDS Accelerator for Apache Spark leverages GPUs to accelerate processing
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:8]:
```suggestion
``Spark-Rapids`` is the RAPIDS Accelerator for Apache Spark, which leverages GPUs to accelerate processing
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+via the `RAPIDS libraries <https://rapids.ai/>`_, for more details, please refer to the
+`spark-rapids <https://nvidia.github.io/spark-rapids/>`_ page.
+
+This tutorial will show you how to use **XGBoost4J-Spark-GPU**.
+
+************************************************
+Build an ML Application with XGBoost4J-Spark-GPU
+************************************************
+
+Refer to XGBoost4J-Spark-GPU related Dependencies
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:18]: This is not a title.

Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:18]:
```suggestion
Refer to XGBoost4J-Spark-GPU for the related Dependencies
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+can define the columns' name as well as their types; otherwise the column name would be
+the default ones derived by Spark, such as ``_col0``, etc. Finally, we can use Spark's
+built-in csv reader to load Iris csv file as a DataFrame named ``xgbInput``.
+
+Spark also contains many built-in readers for other format. eg ORC, Parquet, Avro, Json.
+
+Transform Raw Iris Dataset
+--------------------------
+
+To make Iris dataset be recognizable to XGBoost, we need to transform String-typed
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:69]:
```suggestion
To make Iris dataset be recognizable to XGBoost, we need to encode String-typed
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+label, i.e. "class", to Double-typed label.
+
+To convert String-typed label to Double, we can use Spark's built-in feature transformer
+`StringIndexer <https://spark.apache.org/docs/2.3.1/api/scala/index.html#org.apache.spark.ml.feature.StringIndexer>`_.
+but it has not been accelerated by by Spark-Rapids, which means it will be fallen back
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:74]:
```suggestion
but it has not been accelerated by Spark-Rapids yet, which means it will fall back
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+Spark also contains many built-in readers for other format. eg ORC, Parquet, Avro, Json.
+
+Transform Raw Iris Dataset
+--------------------------
+
+To make Iris dataset be recognizable to XGBoost, we need to transform String-typed
+label, i.e. "class", to Double-typed label.
+
+To convert String-typed label to Double, we can use Spark's built-in feature transformer
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:72]:
```suggestion
One way to convert the String-typed label to Double is to use Spark's built-in feature transformer
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+`StringIndexer <https://spark.apache.org/docs/2.3.1/api/scala/index.html#org.apache.spark.ml.feature.StringIndexer>`_.
+but it has not been accelerated by by Spark-Rapids, which means it will be fallen back
+to CPU to run and cause performance issue. Instead, we use an alternative way to acheive
+this goal by the following code
+
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:77]:
```suggestion
to CPU. As a result, we roll out our own pre-processing step here for greater efficiency:

```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+	|         4.6|        3.2|         1.4|        0.2|    0|
+	+------------+-----------+------------+-----------+-----+
+
+
+With window operations, we have mapped string column of labels to label indices.
+
+Training
+========
+
+XGBoost supports both regression and classification. While we use Iris dataset in
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:110]:
```suggestion
The GPU version of XGBoost-Spark supports both regression and classification models. Although we use the Iris dataset in
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+this tutorial to show how we use XGBoost/XGBoost4J-Spark to resolve a multi-classes
+classification problem, the usage in Regression is very similar to classification.
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:112]:
```suggestion
this tutorial to demonstrate training XGBoost/XGBoost4J-Spark multi-classes
classification model, the concepts introduced should be readily applicable to regression problems as well.
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+
+  val featuresNames = schema.fieldNames.filter(name => name != labelName)
+
+  val xgbClassifier = new XGBoostClassifier(xgbParam)
+      .setFeaturesCol(featuresNames)
+      .setLabelCol(labelName)
+
+The available parameters for training a XGBoost model can be found in :doc:`here </parameter>`.
+Same with XGBoost4J-Spark, XGBoost4J-Spark-GPU also supports not only the default set of parameters
+but also the camel-case variant of these parameters to keep consistent with Spark's MLLIB parameters.
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:134]:
```suggestion
The available parameters for training a XGBoost model can be found in :doc:`here </parameter>`.
Similar to the XGBoost4J-Spark package, in addition to the default set of parameters, XGBoost4J-Spark-GPU also supports the camel-case variant of these parameters to be consistent with Spark's MLLIB naming convention.
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+  val xgbClassifier = new XGBoostClassifier(xgbParam)
+      .setFeaturesCol(featuresNames)
+      .setLabelCol(labelName)
+  xgbClassifier.setMaxDepth(2)
+
+.. note::
+
+  XGBoost4J-Spark first needs to assemble the numeric feature columns into one column with VectorUDF
+  type by VectorAssembler, but XGBoost4J-Spark-GPU does not need such transformation, it accepts an
+  array of feature column names by ``setFeaturesCol(featuresNames)``.
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:152]:
```suggestion
  In contrast to the XGBoost4J-Spark package, which needs to first assemble the numeric feature columns into one column with VectorUDF type by VectorAssembler, the XGBoost4J-Spark-GPU does not require such transformation, it accepts an array of feature column names by ``setFeaturesCol(featuresNames)``.
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+After we set XGBoostClassifier parameters and feature/label columns, we can build a transformer,
+XGBoostClassificationModel by fitting XGBoostClassifier with the input DataFrame. This ``fit``
+operation is essentially the training process and the generated model can then be used in prediction.
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:156]:
```suggestion
operation is essentially the training process and the generated model can then be used in other tasks like prediction.
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+* XGBoostClassificationModel will output margins (``rawPredictionCol``), probabilities(``probabilityCol``) and the eventual prediction labels (``predictionCol``) for each possible label.
+* XGBoostRegressionModel will output prediction label(``predictionCol``).
+
+.. code-block:: scala
+
+  val xgbClassificationModel = xgbClassifier.fit(train)
+  val results = xgbClassificationModel.transform(test)
+  results.show()
+
+With the above code snippet, we get a result DataFrame, result containing margin, probability for each class and the prediction for each instance
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:178]:
```suggestion
With the above code snippet, we get a DataFrame as result, which contains the margin, probability for each class, and the prediction for each instance
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+    --conf spark.task.resource.gpu.amount=0.08 \
+    --conf spark.rapids.sql.enabled=true \
+    --conf spark.rapids.sql.csv.read.double.enabled=true \
+    --conf spark.rapids.sql.hasNans=false\
+    --conf spark.sql.adaptive.enabled=false \
+    --conf spark.rapids.sql.explain=ALL \
+    --conf spark.plugins=com.nvidia.spark.SQLPlugin \
+
+* First, we need to specify the spark-rapids, cudf, xgboost4j-gpu, xgboost4j-spark-gpu packages by ``--packages``
+* Second, ``spark-rapids`` is a Spark plugin, so we need to configure it by ``spark.plugins=com.nvidia.spark.SQLPlugin``
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:234]:
```suggestion
* Second, ``spark-rapids`` is a Spark plugin, so we need to configure it by specifying ``spark.plugins=com.nvidia.spark.SQLPlugin``
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+
+For ``spark.executor.resource.gpu.amount` and `spark.task.resource.gpu.amount``, which is related to GPU scheduling, please refer
+to `Spark GPU Scheduling Overview <https://nvidia.github.io/spark-rapids/Getting-Started/#spark-gpu-scheduling-overview>`_
+
+when enabling `spark.rapids.sql.explain=ALL`, we can get some useful information about whether some spark physical plans can be
```
Comment (trivialfis) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:239]:
```suggestion
when `spark.rapids.sql.explain=ALL` is enabled, we can get some useful information about whether some spark physical plans can be
```

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+    --conf spark.rapids.sql.explain=ALL \
+    --conf spark.plugins=com.nvidia.spark.SQLPlugin \
+
+* First, we need to specify the spark-rapids, cudf, xgboost4j-gpu, xgboost4j-spark-gpu packages by ``--packages``
+* Second, ``spark-rapids`` is a Spark plugin, so we need to configure it by specifying ``spark.plugins=com.nvidia.spark.SQLPlugin``
+
+For ``spark.executor.resource.gpu.amount` and `spark.task.resource.gpu.amount``, which is related to GPU scheduling, please refer
+to `Spark GPU Scheduling Overview <https://nvidia.github.io/spark-rapids/Getting-Started/#spark-gpu-scheduling-overview>`_
+
+when `spark.rapids.sql.explain=ALL` is enabled, we can get some useful information about whether some spark physical plans can be
```
Comment (GaryShen2008) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:241]: Below part from line 241 to line 283 is not related to xgboost, i think.
Shall you remove it?

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+    --conf spark.executor.cores=12 \
+    --conf spark.task.cpus=1 \
+    --conf spark.executor.resource.gpu.amount=1 \
+    --conf spark.task.resource.gpu.amount=0.08 \
+    --conf spark.rapids.sql.enabled=true \
+    --conf spark.rapids.sql.csv.read.double.enabled=true \
+    --conf spark.rapids.sql.hasNans=false\
+    --conf spark.sql.adaptive.enabled=false \
+    --conf spark.rapids.sql.explain=ALL \
+    --conf spark.plugins=com.nvidia.spark.SQLPlugin \
```
Comment (GaryShen2008) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:233]: Shall you provide a complete command here?
The end "\" looks strange.

**Review by GaryShen2008 — CHANGES_REQUESTED** [review]

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+    --class ${main_class}\
+     ${app_jar}\
```
Comment (GaryShen2008) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst]: nit: add a space between the parameter and "\".
And why you need the last "\" if there's no more parameter?

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+  app_jar=iris-1.0.0.jar
+
+  spark-submit \
+    --master $master \
+    --packages ai.rapids:cudf:${cudf_version},com.nvidia:rapids-4-spark_2.12:${rapids_version},ml.dmlc:xgboost4j-gpu_2.12:${xgboost_version},ml.dmlc:xgboost4j-spark-gpu_2.12:${xgboost_version} \
+    --conf spark.executor.cores=12 \
+    --conf spark.task.cpus=1 \
+    --conf spark.executor.resource.gpu.amount=1 \
+    --conf spark.task.resource.gpu.amount=0.08 \
+    --conf spark.rapids.sql.enabled=true \
```
Comment (GaryShen2008) [inline doc/jvm/xgboost4j_spark_gpu_tutorial.rst:238]: Some parameters are not necessary, I think.
Like spark.rapids.sql.enabled, spark.sql.adaptive.enabled, spark.rapids.sql.explain.

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst (hunk 16 above)
Comment (wbo4958) [reply doc/jvm/xgboost4j_spark_gpu_tutorial.rst]: Done

File: doc/jvm/xgboost4j_spark_gpu_tutorial.rst (hunk 17 above)
Comment (wbo4958) [reply doc/jvm/xgboost4j_spark_gpu_tutorial.rst:243]: Done

**Comment by wbo4958** [comment]
@trivialfis could you help merging this PR

**Review by trivialfis — APPROVED** [review]
LGTM
